fdbserver/storageserver.actor.cpp

```diff
@@ -4960,33 +4970,66 @@ ACTOR Future<Void> getMappedKeyValuesQ(StorageServer* data, GetMappedKeyValuesRe
 			req.reply.send(none);
 		} else {
+			state GetMappedKeyValuesReply result;
 			state int remainingLimitBytes = req.limitBytes;
+			state int remainingLimitRows = req.limit;
+			state int pageSize;
+			state bool moreBatch = true;
+			state KeyRef lastIndexKey;
+			state ValueRef lastIndexValue;
+			while (remainingLimitRows > 0 && remainingLimitBytes > 0 && moreBatch) {
+				// do it in a batch way:
+				// 		each batch scans up to PREFETCH_INDEX_PAGE_SIZE indexes
+				// 		fetch the record for each index entry
+
+				// readRange changes the value of its pLimitBytes parameter, while we only want it to be changed later
+				int tmpLimitBytes = req.limitBytes;
+				pageSize = std::min(SERVER_KNOBS->PREFETCH_INDEX_PAGE_SIZE, remainingLimitRows);
+				state GetKeyValuesReply getKeyValuesReply = wait(readRange(data,
+				                                                           version,
+				                                                           KeyRangeRef(begin, end),
+				                                                           pageSize,
+				                                                           &tmpLimitBytes,
+				                                                           span.context,
+				                                                           req.options,
+				                                                           tenantPrefix));
+
+				// Unlock read lock before the subqueries because each
+				// subquery will route back to getValueQ or getKeyValuesQ with a new request having the same
+				// read options which will each acquire the ssLock.
+				readLock.release();
 
-			GetKeyValuesReply getKeyValuesReply = wait(readRange(data,
-			                                                     version,
-			                                                     KeyRangeRef(begin, end),
-			                                                     req.limit,
-			                                                     &remainingLimitBytes,
-			                                                     span.context,
-			                                                     req.options,
-			                                                     tenantPrefix));
-
-			// Unlock read lock before the subqueries because each
-			// subquery will route back to getValueQ or getKeyValuesQ with a new request having the same
-			// read options which will each acquire the ssLock.
-			readLock.release();
-
-			state GetMappedKeyValuesReply r;
-			try {
-				// Map the scanned range to another list of keys and look up.
-				GetMappedKeyValuesReply _r =
-				    wait(mapKeyValues(data, getKeyValuesReply, req.mapper, &req, tenantPrefix, req.matchIndex));
-				r = _r;
-			} catch (Error& e) {
-				TraceEvent("MapError").error(e);
-				throw;
+				// these fields always return as the value of the last batch result
+				result.version = getKeyValuesReply.version;
+				result.cached = getKeyValuesReply.cached;
```

Comment:
What does this `cached` value represent? Is it correct to always set it to the last one?

Reply:
we can remove it, because I do not think it is used anywhere, @ohadzeliger can you please verify that RL has not been using this field?

Reply:
Do you mean `cached` field?

Reply:
oh i mean the value of this field might not used in prefetch code path, not deleting this field from code